shared/src/main/scala/LiterateLexer.scala

```diff
@@ -0,0 +1,158 @@
+package vyxal
+
+import java.util.regex.Pattern
+import scala.util.matching.Regex
+import scala.util.parsing.combinator.*
+import LiterateToken.*
+
+val hardcodedKeywords = Map(
+  // These can't go in the big map, because that's autogenerated
+  "if" -> "[",
+  "endif" -> "}",
+  "end-if" -> "}",
+  "for" -> "(",
+  "endfor" -> "}",
+  "end-for" -> "}",
+  "while" -> "{",
+  "endwhile" -> "}",
+  "end-while" -> "}",
+  "lambda" -> "λ",
+  "endlambda" -> "}",
+  "end-lambda" -> "}",
+  "end" -> "}",
+  "else" -> "|",
+  "do-to-each" -> "(",
+  "body" -> "|",
+  "branch" -> "|",
+  "close-all" -> "]"
+)
+
+enum LiterateToken(val value: Object):
+  // Object instead of String like the normal lexer because it's way easier
+  case Word(override val value: String) extends LiterateToken(value)
+  case AlreadyCode(override val value: String) extends LiterateToken(value)
+  // This is for strings that are already in SBCS form
+  case LitComment(override val value: String) extends LiterateToken(value)
+  case LambdaBlock(override val value: List[Object])
+      extends LiterateToken(value.toString)
+
+  case ListToken(override val value: List[Object])
+      extends LiterateToken(value.toString)
+object LiterateLexer extends RegexParsers:
+  override def skipWhitespace = true
+  override val whiteSpace: Regex = "[ \t\r\f]+".r
+
+  private def decimalRegex = raw"((0|[1-9][0-9]*)?\.[0-9]*|0|[1-9][0-9]*)"
+  def number: Parser[LiterateToken] =
+    raw"($decimalRegex[ij]$decimalRegex?)|([ij]$decimalRegex)|$decimalRegex|([ij]( |$$))".r ^^ {
+      value =>
+        AlreadyCode(
+          value.replace("i", "ı").replace("j", "ı")
+        )
+    }
+
+  def string: Parser[LiterateToken] = raw"""("(?:[^"\\]|\\.)*["])""".r ^^ {
+    value => AlreadyCode(value)
+  }
+
+  def singleCharString: Parser[LiterateToken] = """'.""".r ^^ { value =>
+    AlreadyCode(value)
+  }
+
+  def comment: Parser[LiterateToken] = """##[^\n]*""".r ^^ { value =>
+    LitComment(value)
+  }
+
+  def lambdaBlock: Parser[LiterateToken] =
+    """\{""".r ~ rep(lambdaBlock | """(#}|[^{}])+""".r) ~ """\}""".r ^^ {
+      case _ ~ body ~ _ => LambdaBlock(body)
+    }
+  def normalGroup: Parser[LiterateToken] =
+    """\(""".r ~ rep(normalGroup | """[^()]+""".r) ~ """\)""".r ^^ {
+      case _ ~ body ~ _ => Word(sbcsify(body.map(recHelp).mkString))
+    }
+
+  def list: Parser[LiterateToken] =
+    "[" ~ repsep(list | """[^\]\[|]+""".r, "|") ~ "]" ^^ { case _ ~ body ~ _ =>
```

Comment:
```suggestion
    "[" ~> repsep(list | """[^\]\[|]+""".r, "|") <~ "]" ^^ { body =>
```
Since we don't care about the []